pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,1508 @@
+			// The external filter does filter for user as, possible
+			// conflict.
+			if callerQuest[UserIDLabel] != user {
+				// It asks for a user other than the current.
+				// We can bail now, with an empty result, as
+				// nothing can match.
+				return localOptions, nil
+			}
+			// It asks for the current user, same as the internal
+			// filter would do.  We can pass the options as is.
+		} else {
+			// The external filter has nothing about the user.
+			// We can simply add the internal filter.
+			localOptions.LabelSelector = labels.Merge(callerQuest, quest).AsSelector().String()
+		}
+	}
+
+	return localOptions, nil
+}
+
+// secretFromToken converts the token argument into the equivalent secrets to
+// store in k8s.
+func secretFromToken(token *ext.Token, oldBackendLabels, oldBackendAnnotations map[string]string) (*corev1.Secret, error) {
+	// inject default on creation
+	ttl := token.Spec.TTL
+	if ttl == 0 {
+		ttl = ThirtyDays
+		// pass back to caller (Create)
+		token.Spec.TTL = ttl
+	}
+
+	// annotations -- encode and store into a data field. on reading decode
+	// that field.  this fully separates the annotations on the tokens from
+	// the annotations on the backing secrets.
+
+	annotationBytes, err := json.Marshal(token.Annotations)
+	if err != nil {
+		return nil, err
+	}
+
+	// labels -- encode and store into a data field. on reading decode that
+	// field.  this fully separates the user-visible labels on the tokens
+	// from the labels on the backing secrets.
+
+	labelBytes, err := json.Marshal(token.Labels)
+	if err != nil {
+		return nil, err
+	}
+
+	// labels again -- for proper filtering and searching the referenced
+	// user id, and the kind of the token are placed into labels of the
+	// backing secret -- the keys for these labels are part of the public
+	// API. This may have to be merged with labels set on the secrets by
+	// other apps with access to the secrets.
+
+	backendLabels := oldBackendLabels
+	if backendLabels == nil {
+		backendLabels = map[string]string{}
+	}
+	backendLabels[UserIDLabel] = token.Spec.UserID
+	backendLabels[KindLabel] = token.Spec.Kind
+
+	// ensure that only one of name or generateName is passed through.
+	name := token.Name
+	genName := token.GenerateName
+	if genName != "" {
+		name = ""
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    TokenNamespace,
+			Name:         name,
+			GenerateName: genName,
+			Labels:       backendLabels,
+			Annotations:  oldBackendAnnotations,
+		},
+		StringData: make(map[string]string),
+		Data:       make(map[string][]byte),
+	}
+
+	// system
+	secret.StringData[FieldUID] = string(token.ObjectMeta.UID)
+	secret.StringData[FieldLabels] = string(labelBytes)
+	secret.StringData[FieldAnnotations] = string(annotationBytes)
+
+	// spec
+	enabled := token.Spec.Enabled == nil || *token.Spec.Enabled
+
+	secret.StringData[FieldUserID] = token.Spec.UserID
+	secret.StringData[FieldClusterName] = token.Spec.ClusterName
```

Comment:
If we allow creating a cluster-scoped ext tokens right away how do we handle the syncing it with ClusterAuthToken's in downstreams? Note that we have bidirectional syncing between Tokens and ClusterAuthTokens.

This is also brings a bigger question what are we doing about ClusterAuthTokens? Are we keeping them as-is as CRDs that holds the tokens' hash?

Reply:
I have not done anything about ClusterAuthTokens in the context of ext tokens, so far.